pkg/gatewayserver/io/basicstationlns/messages/upstream.go

```diff
@@ -0,0 +1,294 @@
+			AntennaIndex:       uint32(req.RadioMetaData.UpInfo.RCtx),
+		},
+		Timestamp: timestamp,
+	}
+	ulTokenBytes, err := ulToken.Marshal()
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	rxTime := time.Unix(req.RadioMetaData.UpInfo.RxTime, 0)
+	rxMetadata := &ttnpb.RxMetadata{
+		GatewayIdentifiers: ids,
+		Time:               &rxTime,
+		Timestamp:          timestamp,
+		RSSI:               req.RadioMetaData.UpInfo.RSSI,
+		SNR:                req.RadioMetaData.UpInfo.SNR,
+		UplinkToken:        ulTokenBytes,
+	}
+	up.RxMetadata = append(up.RxMetadata, rxMetadata)
+
+	loraDR, err := getDataRateFromIndex(bandID, req.RadioMetaData.DataRate)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+	up.Settings = ttnpb.TxSettings{
+		Frequency:     req.RadioMetaData.Frequency,
+		DataRateIndex: ttnpb.DataRateIndex(req.RadioMetaData.DataRate),
+		DataRate:      loraDR,
+	}
+
+	return up, nil
+}
+
+// ToUplinkMessage extracts fields from the basic station Join Request "jreq" message and converts them into an UplinkMessage for the network server.
+func (updf *UplinkDataFrame) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (ttnpb.UplinkMessage, error) {
+	up := ttnpb.UplinkMessage{}
+	up.ReceivedAt = time.Now()
+
+	parsedMHDR := ttnpb.MHDR{}
+	err := lorawan.UnmarshalMHDR([]byte{byte(updf.MHdr)}, &parsedMHDR)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	if (parsedMHDR.MType != ttnpb.MType_UNCONFIRMED_UP) && (parsedMHDR.MType != ttnpb.MType_CONFIRMED_UP) {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame
+	}
+
+	micBytes, err := getInt32AsByteSlice(updf.MIC)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	var fport uint32
+	if updf.FPort == -1 {
+		fport = 0
+	} else {
+		fport = uint32(updf.FPort)
+	}
+
+	devAddrBytes, err := getInt32AsByteSlice(updf.DevAddr)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	devAddr := types.DevAddr{}
+	err = devAddr.UnmarshalBinary(devAddrBytes)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	fctrl := ttnpb.FCtrl{}
+	err = lorawan.UnmarshalFCtrl([]byte{byte(updf.FCtrl)}, &fctrl, true)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	up.Payload = &ttnpb.Message{
+		MHDR: parsedMHDR,
+		MIC:  micBytes,
+		Payload: &ttnpb.Message_MACPayload{MACPayload: &ttnpb.MACPayload{
+			FPort:      fport,
+			FRMPayload: []byte(updf.FRMPayload),
+			FHDR: ttnpb.FHDR{
+				DevAddr: devAddr,
+				FCtrl:   fctrl,
+				FCnt:    uint32(updf.FCnt),
+				FOpts:   []byte(updf.FOpts),
+			},
+		}},
+	}
+
+	up.RawPayload, err = lorawan.MarshalMessage(*up.Payload)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	timestamp := uint32(updf.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
```

Comment:
Not sure if this goes well with the sign; please test this

Reply:
`XTime` cannot be greater than 32 bits until 2038 and should not be represented as a negative number. So just extracting the last 32 bits should not be an issue afaik.

Reply:
I see; it's a `int64`; I thought `int32`.

Docs say;

>[...] and `xtime` are passed along through the LNS and MUST be added to matching Class A downlink messages.

Is this going to work if we only pass 32 lsb?

Reply:
Well we discussed that we are going to generate the `xtime` for the Downlink based on the `time`/`timestamp` of `TxSettings`. Passing the original `xtime` will override the GS scheduling.